File: src/erenshor/application/wiki/generators/sections/spell.py
"""Spell section generator for wiki content.

This module generates MediaWiki {{Ability}} template wikitext for spell entities.

This section generator produces templates for single spells. Multi-entity page
assembly is handled by PageGenerator classes.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from loguru import logger

from erenshor.application.wiki.generators.formatting import format_description, safe_str
from erenshor.application.wiki.generators.sections.base import SectionGeneratorBase

if TYPE_CHECKING:
    from erenshor.application.wiki.services.class_display_service import ClassDisplayNameService
    from erenshor.domain.enriched_data.spell import EnrichedSpellData

# Game constants for cast time calculation
GAME_TICKS_PER_SECOND = 60  # Game runs at 60 ticks per second


class SpellSectionGenerator(SectionGeneratorBase):
    """Generator for spell wiki sections.

    Generates {{Ability}} template wikitext for a single spell entity.

    Multi-entity page assembly is handled by PageGenerator classes, not here.
    """

    def __init__(self, class_display: ClassDisplayNameService) -> None:
        """Initialize spell template generator.

        Args:
            class_display: Service for mapping class names to display names
        """
        super().__init__()
        self._class_display = class_display

    def generate_template(
        self,
        enriched: EnrichedSpellData,
        page_title: str,
    ) -> str:
        """Generate {{Ability}} template wikitext for a single spell."""
        spell = enriched.spell
        logger.debug(f"Generating template for spell: {spell.spell_name}")

        context = self._build_spell_template_context(enriched, page_title)
        template_wikitext = self.render_template("ability.jinja2", context)
        return self.normalize_wikitext(template_wikitext)

    def _build_spell_template_context(
        self,
        enriched: EnrichedSpellData,
        page_title: str,
    ) -> dict[str, str]:
        """Build context for {{Ability}} template from Spell entity."""
        spell = enriched.spell

        def bool_str(value: int | None) -> str:
            return "True" if value else ""

        # Calculate duration from ticks (if available)
        duration = ""
        has_duration = bool(spell.spell_duration_in_ticks)
        if spell.spell_duration_in_ticks:
            duration = self._format_duration(spell.spell_duration_in_ticks)

        # Format class restrictions with level: [[DisplayName]] (level)
        classes_list = []
        if enriched.classes and spell.required_level and spell.required_level > 0:
            display_names = self._class_display.map_class_list(enriched.classes)
            for display_name in display_names:
                classes_list.append(f"[[{display_name}]] ({spell.required_level})")
        classes = "<br>".join(classes_list)

        cast_time_str = self._format_cast_time(spell.spell_charge_time)

        image = f"{spell.image_name}.png" if spell.image_name else ""

        # Pre-built links on the Spell entity and enriched DTO
        pet_to_summon = str(enriched.pet_to_summon) if enriched.pet_to_summon else ""
        status_effect = str(spell.status_effect_link) if spell.status_effect_link else ""
        add_proc = str(spell.add_proc_link) if spell.add_proc_link else ""

        imagecaption = ""
        if spell.status_effect_message_on_player:
            imagecaption = f"You {spell.status_effect_message_on_player}"

        display_name = spell.display_name or spell.spell_name or page_title

        # Items with effect and teaching items are already ItemLink objects
        itemswitheffect = self._format_wiki_links(enriched.items_with_effect)
        source = self._format_wiki_links(enriched.teaching_items)
        used_by = self._format_wiki_links(enriched.used_by_characters)

        context: dict[str, str] = {
            "title": display_name,
            "image": image,
            "imagecaption": imagecaption,
            "description": format_description(safe_str(spell.spell_desc)) if spell.spell_desc else "",
            "type": safe_str(spell.type),
            "line": safe_str(spell.line),
            "classes": classes,
            "required_level": safe_str(spell.required_level, zero_as_blank=True),
            "manacost": safe_str(spell.mana_cost, zero_as_blank=True),
            "aggro": safe_str(spell.aggro, zero_as_blank=True),
            "is_taunt": bool_str(spell.taunt_spell),
            "casttime": cast_time_str,
            "cooldown": self._format_cooldown(spell.cooldown),
            "duration": duration,
            "duration_in_ticks": safe_str(spell.spell_duration_in_ticks, zero_as_blank=True),
            "has_unstable_duration": bool_str(spell.unstable_duration),
            "is_instant_effect": bool_str(spell.instant_effect),
            "is_reap_and_renew": bool_str(spell.reap_and_renew),
            "is_sim_usable": bool_str(spell.sim_usable),
            "range": "" if spell.self_only else (str(int(spell.spell_range)) if spell.spell_range else ""),
            "max_level_target": safe_str(spell.max_level_target, zero_as_blank=True),
            "is_self_only": bool_str(spell.self_only),
            "is_group_effect": bool_str(spell.group_effect),
            "is_applied_to_caster": bool_str(spell.apply_to_caster),
            "effects": "",  # TODO: Build from multiple effect fields
            "damage_type": safe_str(spell.damage_type) if spell.target_damage else "",
            "resist_modifier": str(int(spell.resist_modifier)) if spell.resist_modifier else "",
            "target_damage": safe_str(spell.target_damage, zero_as_blank=True),
            "target_healing": safe_str(spell.target_healing, zero_as_blank=True),
            "caster_healing": safe_str(spell.caster_healing, zero_as_blank=True),
            "shield_amount": safe_str(spell.shielding_amt, zero_as_blank=True),
            "pet_to_summon": pet_to_summon,
            "status_effect": status_effect,
            "add_proc": add_proc,
            "add_proc_chance": safe_str(spell.add_proc_chance, zero_as_blank=True),
            "has_lifetap": bool_str(spell.lifetap),
            "lifesteal": safe_str(spell.percent_lifesteal, zero_as_blank=True),
            "damage_shield": safe_str(spell.damage_shield, zero_as_blank=True),
            "percent_mana_restoration": safe_str(spell.percent_mana_restoration, zero_as_blank=True),
            "bleed_damage_percent": safe_str(spell.bleed_damage_percent, zero_as_blank=True),
            "special_descriptor": safe_str(spell.special_descriptor),
            # Stat modifiers
            "hp": safe_str(spell.hp, zero_as_blank=True),
            "ac": safe_str(spell.ac, zero_as_blank=True),
            "mana": safe_str(spell.mana, zero_as_blank=True),
            "str": safe_str(spell.str_, zero_as_blank=True),
            "dex": safe_str(spell.dex, zero_as_blank=True),
            "end": safe_str(spell.end_, zero_as_blank=True),
            "agi": safe_str(spell.agi, zero_as_blank=True),
            "wis": safe_str(spell.wis, zero_as_blank=True),
            "int": safe_str(spell.int_, zero_as_blank=True),
            "cha": safe_str(spell.cha, zero_as_blank=True),
            "mr": safe_str(spell.mr, zero_as_blank=True),
            "er": safe_str(spell.er, zero_as_blank=True),
            "vr": safe_str(spell.vr, zero_as_blank=True),
            "pr": safe_str(spell.pr, zero_as_blank=True),
            "haste": safe_str(spell.haste, zero_as_blank=True),
            "resonance": safe_str(spell.resonate_chance, zero_as_blank=True),
            "movement_speed": safe_str(spell.movement_speed, zero_as_blank=True),
            "atk_roll_modifier": safe_str(spell.atk_roll_modifier, zero_as_blank=True),
            "xp_bonus": safe_str(spell.xp_bonus, zero_as_blank=True) if has_duration else "",
            # Crowd control
            "is_root": bool_str(spell.root_target),
            "is_stun": bool_str(spell.stun_target),
            "is_charm": bool_str(spell.charm_target),
            "is_broken_on_damage": bool_str(spell.break_on_damage),
            "is_fear": bool_str(spell.fear_target),
            "inflict_on_self": bool_str(spell.inflict_on_self),
            # Sources
            "itemswitheffect": itemswitheffect,
            "source": source,
            "used_by": used_by,
        }

        return context

    def _format_cast_time(self, spell_charge_time: float | None) -> str:
        """Format spell cast time in ticks to human-readable string."""
        if spell_charge_time is None or spell_charge_time == 0:
            return "Instant"

        seconds = spell_charge_time / GAME_TICKS_PER_SECOND

        if seconds < 0.05:
            return "Instant"

        return f"{seconds:.1f} seconds"

    def _format_duration(self, duration_in_ticks: int) -> str:
        """Format spell duration from ticks to human-readable string."""
        seconds = duration_in_ticks * 3
        return f"{seconds} seconds"

    def _format_cooldown(self, cooldown: float | None) -> str:
        """Format spell cooldown to human-readable string."""
        if cooldown is None or cooldown == 0:
            return ""

        return f"{int(cooldown)} seconds"

    def _format_wiki_links(self, links: list) -> str:  # type: ignore[type-arg]
        """Format a list of WikiLink objects as wikitext separated by <br>.

        Filters out links with no page_title (excluded entities), sorts by display
        name, and joins with <br>.
        """
        if not links:
            return ""

        visible = [link for link in links if link.page_title is not None]
        visible.sort()
        return "<br>".join(str(link) for link in visible)

File: wiki/modules/Erenshor/Spell.lua
local AbilityLink = require("Module:Erenshor/AbilityLink")
local Args = require("Module:Erenshor/Args")
local Link = require("Module:Erenshor/Link")
local Format = require("Module:Erenshor/Format")
local Tooltip = require("Module:Erenshor/Spell/Tooltip")
local Cargo = require("Module:Erenshor/Cargo")

local Data = mw.loadData("Module:Erenshor/Data/Spells")

local p = {}

local FIELD_OVERRIDES = {
	["title"] = "name",
	["image"] = "image",
	["imagecaption"] = "imageCaption",
	["description"] = "description",
	["type"] = "type",
	["line"] = "line",
	["classes"] = "classesOverride",
	["required_level"] = "requiredLevel",
	["manacost"] = "manaCost",
	["aggro"] = "aggro",
	["is_taunt"] = "taunt",
	["casttime"] = "castTimeOverride",
	["cooldown"] = "cooldownOverride",
	["duration"] = "durationOverride",
	["has_unstable_duration"] = "unstableDuration",
	["is_instant_effect"] = "instantEffect",
	["is_reap_and_renew"] = "reapAndRenew",
	["is_sim_usable"] = "simUsable",
	["range"] = "range",
	["max_level_target"] = "maxLevelTarget",
	["is_self_only"] = "selfOnly",
	["is_group_effect"] = "groupEffect",
	["is_applied_to_caster"] = "applyToCaster",
	["effects"] = "effects",
	["damage_type"] = "damageType",
	["resist_modifier"] = "resistModifier",
	["target_damage"] = "targetDamage",
	["target_healing"] = "targetHealing",
	["caster_healing"] = "casterHealing",
	["shield_amount"] = "shieldAmount",
	["pet_to_summon"] = "petToSummon",
	["status_effect"] = "statusEffectOverride",
	["add_proc"] = "addProcOverride",
	["add_proc_chance"] = "addProcChance",
	["has_lifetap"] = "lifetap",
	["lifesteal"] = "lifesteal",
	["damage_shield"] = "damageShield",
	["percent_mana_restoration"] = "percentManaRestoration",
	["bleed_damage_percent"] = "bleedDamagePercent",
	["special_descriptor"] = "specialDescriptor",
	["hp"] = "hp",
	["ac"] = "ac",
	["mana"] = "mana",
	["str"] = "str",
	["dex"] = "dex",
	["end"] = "end",
	["agi"] = "agi",
	["wis"] = "wis",
	["int"] = "int",
	["cha"] = "cha",
	["mr"] = "mr",
	["er"] = "er",
	["vr"] = "vr",
	["pr"] = "pr",
	["haste"] = "haste",
	["resonance"] = "resonance",
	["movement_speed"] = "movementSpeed",
	["atk_roll_modifier"] = "atkRollModifier",
	["xp_bonus"] = "xpBonus",
	["is_root"] = "root",
	["is_stun"] = "stun",
	["is_charm"] = "charm",
	["is_broken_on_damage"] = "breakOnDamage",
	["is_fear"] = "fear",
	["inflict_on_self"] = "inflictOnSelf",
	["itemswitheffect"] = "itemsWithEffect",
	["source"] = "source",
	["used_by"] = "usedBy",
}

local ROOT_PUBLIC_PARAMETERS = {
	"title",
	"image",
	"imagecaption",
	"description",
	"type",
	"line",
	"classes",
	"required_level",
	"manacost",
	"aggro",
	"is_taunt",
	"casttime",
	"cooldown",
	"duration",
	"has_unstable_duration",
	"is_instant_effect",
	"is_reap_and_renew",
	"is_sim_usable",
	"range",
	"max_level_target",
	"is_self_only",
	"is_group_effect",
	"is_applied_to_caster",
	"effects",
	"damage_type",
	"resist_modifier",
	"target_damage",
	"target_healing",
	"caster_healing",
	"shield_amount",
	"pet_to_summon",
	"status_effect",
	"add_proc",
	"add_proc_chance",
	"has_lifetap",
	"lifesteal",
	"damage_shield",
	"percent_mana_restoration",
	"bleed_damage_percent",
	"special_descriptor",
	"hp",
	"ac",
	"mana",
	"str",
	"dex",
	"end",
	"agi",
	"wis",
	"int",
	"cha",
	"mr",
	"er",
	"vr",
	"pr",
	"haste",
	"resonance",
	"movement_speed",
	"atk_roll_modifier",
	"xp_bonus",
	"is_root",
	"is_stun",
	"is_charm",
	"is_broken_on_damage",
	"is_fear",
	"inflict_on_self",
	"itemswitheffect",
	"source",
	"used_by",
}

local function copyTable(value)
	local out = {}
	if value == nil then
		return out
	end
	for key, item in pairs(value) do
		if type(item) == "table" then
			out[key] = copyTable(item)
		else
			out[key] = item
		end
	end
	return out
end

local function templateArgs(frame)
	local out = copyTable(Args.parentArgs(frame))
	if frame ~= nil and frame.args ~= nil then
		for key, value in pairs(frame.args) do
			out[key] = value
		end
	end
	return out
end

local function currentTitleText()
	if mw ~= nil and mw.title ~= nil and mw.title.getCurrentTitle ~= nil then
		return mw.title.getCurrentTitle().text
	end
	return ""
end

local function isBlank(value)
	return value == nil or tostring(value):match("^%s*$") ~= nil
end

local function ensureImageFile(image, fallbackName)
	local value = image
	if isBlank(value) then
		value = fallbackName
	end
	if isBlank(value) then
		return nil
	end
	value = tostring(value)
	if value:match("%.[Pp][Nn][Gg]$") or value:match("%.[Jj][Pp][Gg]$") then
		return value
	end
	return value .. ".png"
end

local function numberText(value)
	local number = tonumber(value)
	if number == nil then
		return tostring(value)
	end
	if number == math.floor(number) then
		return tostring(math.floor(number))
	end
	return tostring(number)
end

local function nonZeroNumberText(value)
	local number = tonumber(value)
	if number == nil or number == 0 then
		return ""
	end
	return numberText(number)
end

local function boolText(value)
	if value == true or value == "1" or value == "True" or value == "true" then
		return "True"
	end
	return ""
end

local function secondsText(value)
	local number = tonumber(value)
	if number == nil or number == 0 then
		return ""
	end
	return numberText(number) .. " seconds"
end

local function castTimeText(spell)
	if not isBlank(spell.castTimeOverride) then
		return spell.castTimeOverride
	end
	local seconds = tonumber(spell.castTimeSeconds)
	if seconds == nil or seconds < 0.05 then
		return "Instant"
	end
	return string.format("%.1f seconds", seconds)
end

local function cooldownText(spell)
	if not isBlank(spell.cooldownOverride) then
		return spell.cooldownOverride
	end
	return secondsText(spell.cooldownSeconds)
end

local function durationText(spell)
	if not isBlank(spell.durationOverride) then
		return spell.durationOverride
	end
	local seconds = tonumber(spell.durationSeconds)
	if seconds == nil or seconds == 0 then
		return ""
	end
	return numberText(seconds) .. " seconds"
end

local function imageCaptionText(spell)
	if not isBlank(spell.imageCaption) then
		return spell.imageCaption
	end
	if isBlank(spell.statusEffectMessageOnPlayer) then
		return ""
	end
	return "You " .. tostring(spell.statusEffectMessageOnPlayer)
end

local function classesText(spell)
	if not isBlank(spell.classesOverride) then
		return spell.classesOverride
	end
	local level = tonumber(spell.requiredLevel)
	if level == nil or level <= 0 or spell.classes == nil then
		return ""
	end
	local out = {}
	local classLinks = spell.classLinks
	for index, className in ipairs(spell.classes) do
		if not isBlank(className) then
			local classLink = type(classLinks) == "table" and classLinks[index] or nil
			local args
			if type(classLink) == "table" and not isBlank(classLink.stablekey) then
				args = { kind = "class", stablekey = classLink.stablekey }
			else
				args = { kind = "class", page = className }
			end
			table.insert(out, Link.render(args) .. " (" .. numberText(level) .. ")")
		end
	end
	return table.concat(out, "<br>")
end

local function rangeText(spell)
	if spell.selfOnly == true then
		return ""
	end
	return nonZeroNumberText(spell.range)
end

local function lineList(values)
	if values == nil then
		return nil
	end
	if type(values) ~= "table" then
		return values
	end
	local out = {}
	for _, value in ipairs(values) do
		if type(value) == "table" and value.kind ~= nil then
			table.insert(out, Link.render(value))
		else
			table.insert(out, value)
		end
	end
	return table.concat(out, "<br>")
end

local function linkedCharacter(stableKey)
	if isBlank(stableKey) then
		return ""
	end
	return Link.render({ kind = "character", stablekey = stableKey })
end

local function linkedSpell(stableKey)
	if isBlank(stableKey) or Data.spells[stableKey] == nil then
		return ""
	end
	return AbilityLink.render({ stablekey = stableKey })
end

local function statusEffectText(spell)
	if not isBlank(spell.statusEffectOverride) then
		return spell.statusEffectOverride
	end
	return linkedSpell(spell.statusEffectStableKey)
end

local function addProcText(spell)
	if not isBlank(spell.addProcOverride) then
		return spell.addProcOverride
	end
	return linkedSpell(spell.addProcStableKey)
end

local FIELD_ACCESSORS = {
	title = function(s)
		return s.name
	end,
	image = function(s)
		return ensureImageFile(s.image, s.name)
	end,
	imagecaption = imageCaptionText,
	description = function(s)
		return s.description
	end,
	type = function(s)
		return s.type
	end,
	line = function(s)
		return s.line
	end,
	classes = classesText,
	required_level = function(s)
		return nonZeroNumberText(s.requiredLevel)
	end,
	manacost = function(s)
		return nonZeroNumberText(s.manaCost)
	end,
	aggro = function(s)
		return nonZeroNumberText(s.aggro)
	end,
	is_taunt = function(s)
		return boolText(s.taunt)
	end,
	casttime = castTimeText,
	cooldown = cooldownText,
	duration = durationText,
	has_unstable_duration = function(s)
		return boolText(s.unstableDuration)
	end,
	is_instant_effect = function(s)
		return boolText(s.instantEffect)
	end,
	is_reap_and_renew = function(s)
		return boolText(s.reapAndRenew)
	end,
	is_sim_usable = function(s)
		return boolText(s.simUsable)
	end,
	range = rangeText,
	max_level_target = function(s)
		return nonZeroNumberText(s.maxLevelTarget)
	end,
	is_self_only = function(s)
		return boolText(s.selfOnly)
	end,
	is_group_effect = function(s)
		return boolText(s.groupEffect)
	end,
	is_applied_to_caster = function(s)
		return boolText(s.applyToCaster)
	end,
	effects = function(s)
		return s.effects
	end,
	damage_type = function(s)
		if tonumber(s.targetDamage) == nil or tonumber(s.targetDamage) == 0 then
			return ""
		end
		return s.damageType
	end,
	resist_modifier = function(s)
		return nonZeroNumberText(s.resistModifier)
	end,
	target_damage = function(s)
		return nonZeroNumberText(s.targetDamage)
	end,
	target_healing = function(s)
		return nonZeroNumberText(s.targetHealing)
	end,
	caster_healing = function(s)
		return nonZeroNumberText(s.casterHealing)
	end,
	shield_amount = function(s)
		return nonZeroNumberText(s.shieldAmount)
	end,
	pet_to_summon = function(s)
		if not isBlank(s.petToSummon) then
			return s.petToSummon
		end
		return linkedCharacter(s.petToSummonStableKey)
	end,
	status_effect = statusEffectText,
	add_proc = addProcText,
	add_proc_chance = function(s)
		return nonZeroNumberText(s.addProcChance)
	end,
	has_lifetap = function(s)
		return boolText(s.lifetap)
	end,
	lifesteal = function(s)
		return nonZeroNumberText(s.lifesteal)
	end,
	damage_shield = function(s)
		return nonZeroNumberText(s.damageShield)
	end,
	percent_mana_restoration = function(s)
		return nonZeroNumberText(s.percentManaRestoration)
	end,
	bleed_damage_percent = function(s)
		return nonZeroNumberText(s.bleedDamagePercent)
	end,
	special_descriptor = function(s)
		return s.specialDescriptor
	end,
	hp = function(s)
		return nonZeroNumberText(s.hp)
	end,
	ac = function(s)
		return nonZeroNumberText(s.ac)
	end,
	mana = function(s)
		return nonZeroNumberText(s.mana)
	end,
	str = function(s)
		return nonZeroNumberText(s.str)
	end,
	dex = function(s)
		return nonZeroNumberText(s.dex)
	end,
	["end"] = function(s)
		return nonZeroNumberText(s["end"])
	end,
	agi = function(s)
		return nonZeroNumberText(s.agi)
	end,
	wis = function(s)
		return nonZeroNumberText(s.wis)
	end,
	["int"] = function(s)
		return nonZeroNumberText(s["int"])
	end,
	cha = function(s)
		return nonZeroNumberText(s.cha)
	end,
	mr = function(s)
		return nonZeroNumberText(s.mr)
	end,
	er = function(s)
		return nonZeroNumberText(s.er)
	end,
	vr = function(s)
		return nonZeroNumberText(s.vr)
	end,
	pr = function(s)
		return nonZeroNumberText(s.pr)
	end,
	haste = function(s)
		return nonZeroNumberText(s.haste)
	end,
	resonance = function(s)
		return nonZeroNumberText(s.resonance)
	end,
	movement_speed = function(s)
		return nonZeroNumberText(s.movementSpeed)
	end,
	atk_roll_modifier = function(s)
		return nonZeroNumberText(s.atkRollModifier)
	end,
	xp_bonus = function(s)
		return nonZeroNumberText(s.xpBonus)
	end,
	is_root = function(s)
		return boolText(s.root)
	end,
	is_stun = function(s)
		return boolText(s.stun)
	end,
	is_charm = function(s)
		return boolText(s.charm)
	end,
	is_broken_on_damage = function(s)
		return boolText(s.breakOnDamage)
	end,
	is_fear = function(s)
		return boolText(s.fear)
	end,
	inflict_on_self = function(s)
		return boolText(s.inflictOnSelf)
	end,
	itemswitheffect = function(s)
		return lineList(s.itemsWithEffect)
	end,
	source = function(s)
		return lineList(s.source)
	end,
	used_by = function(s)
		return lineList(s.usedBy)
	end,
}

local function explicitStableKey(args)
	return Args.resolve(args, "stablekey", nil)
		or Args.resolve(args, "stableKey", nil)
		or Args.resolve(args, "key", nil)
		or Args.resolve(args, "id", nil)
end

local function resolveStableKey(args)
	local stableKey = explicitStableKey(args)
	if stableKey ~= nil and Data.spells[stableKey] ~= nil then
		return stableKey
	end
	return nil
end

local function applyRootOverrides(spell, args)
	for _, publicName in ipairs(ROOT_PUBLIC_PARAMETERS) do
		local fieldName = FIELD_OVERRIDES[publicName]
		if fieldName ~= nil and Args.has(args, publicName) then
			spell[fieldName] = Args.resolve(args, publicName, spell[fieldName])
		end
	end
end

local function missingSpell(args, pageTitle)
	return {
		missing = true,
		name = Args.resolve(args, "title", pageTitle) or pageTitle,
		page = pageTitle,
	}
end

function p.resolve(args, pageTitle)
	args = args or {}
	pageTitle = pageTitle or currentTitleText()

	local stableKey = resolveStableKey(args)
	if stableKey == nil then
		return missingSpell(args, pageTitle)
	end

	local spell = copyTable(Data.spells[stableKey])
	spell.stableKey = stableKey
	applyRootOverrides(spell, args)
	return spell
end

local function missingOutput(spell)
	return '<span class="erenshor-missing-data">Missing spell data: '
		.. Format.escape(spell.name)
		.. "</span>[[Category:Pages with missing Erenshor spell data]]"
end

function p.fieldValue(args, pageTitle, key)
	local spell = p.resolve(args, pageTitle)
	if spell.missing then
		return ""
	end
	local accessor = FIELD_ACCESSORS[key]
	if accessor == nil then
		return ""
	end
	local value = accessor(spell)
	if value == nil or value == false then
		return ""
	end
	return tostring(value)
end

function p.statusText(args, pageTitle)
	local spell = p.resolve(args, pageTitle)
	if spell.missing then
		return missingOutput(spell)
	end
	return ""
end

local function cargoFields(spell, pageTitle)
	return {
		{ "StableKey", spell.stableKey },
		{ "Page", pageTitle },
		{ "Name", spell.name },
		{ "Image", ensureImageFile(spell.image, spell.name) },
		{ "Type", spell.type },
		{ "Line", spell.line },
		{ "RequiredLevel", spell.requiredLevel },
		{ "ManaCost", spell.manaCost },
		{ "CastTimeSeconds", spell.castTimeSeconds },
		{ "CooldownSeconds", spell.cooldownSeconds },
		{ "DurationSeconds", spell.durationSeconds },
		{ "CastRange", spell.range },
		{ "DamageType", spell.damageType },
		{ "TargetDamage", spell.targetDamage },
		{ "TargetHealing", spell.targetHealing },
		{ "CasterHealing", spell.casterHealing },
		{ "ShieldingAmt", spell.shieldAmount },
		{ "Aggro", spell.aggro },
		{ "SimUsable", spell.simUsable },
		{ "SimsNeedHelpToLearn", spell.simsNeedHelpToLearn },
		{ "SelfOnly", spell.selfOnly },
		{ "GroupEffect", spell.groupEffect },
		{ "CrowdControl", spell.crowdControl },
		{ "GrantInvisibility", spell.grantInvisibility },
		{ "CannotInterrupt", spell.cannotInterrupt },
		{ "Jolt", spell.jolt },
		{ "NoResonate", spell.noResonate },
		{ "ArmorPenPercent", spell.armorPenPercent },
		{ "LevelScaledManaRestoration", spell.levelScaledManaRestoration },
		{ "ShapeshiftForm", spell.shapeshiftForm },
		{ "StatusEffectKey", spell.statusEffectStableKey },
		{ "AddProcKey", spell.addProcStableKey },
		{ "PetToSummonKey", spell.petToSummonStableKey },
	}
end

local function classFields(spell, className)
	return {
		{ "AbilityKey", spell.stableKey },
		{ "Class", className },
		{ "RequiredLevel", spell.requiredLevel },
	}
end

-- A spell's classes are a flat name list; each becomes one AbilityClasses row that
-- broadcasts the spell's single requiredLevel.
local function eachClass(spell, callback)
	if type(spell.classes) ~= "table" then
		return
	end
	for _, className in ipairs(spell.classes) do
		if not isBlank(className) then
			callback(className)
		end
	end
end

function p.field(frame)
	return p.fieldValue(templateArgs(frame), currentTitleText(), frame.args[1])
end

function p.status(frame)
	return p.statusText(templateArgs(frame), currentTitleText())
end

function p.cargoArgs(frame)
	local pageTitle = currentTitleText()
	local spell = p.resolve(templateArgs(frame), pageTitle)
	if spell.missing then
		return {}
	end
	return Cargo.buildArgs("Spells", cargoFields(spell, pageTitle))
end

function p.cargoClassRows(frame)
	local spell = p.resolve(templateArgs(frame), currentTitleText())
	local rows = {}
	if not spell.missing then
		eachClass(spell, function(className)
			table.insert(rows, Cargo.buildArgs("AbilityClasses", classFields(spell, className)))
		end)
	end
	return rows
end

function p.cargoStore(frame)
	local pageTitle = currentTitleText()
	local spell = p.resolve(templateArgs(frame), pageTitle)
	if spell.missing then
		return ""
	end
	Cargo.store("Spells", cargoFields(spell, pageTitle))
	eachClass(spell, function(className)
		Cargo.store("AbilityClasses", classFields(spell, className))
	end)
	return ""
end

function p.renderTooltip(args, pageTitle)
	local spell = p.resolve(args, pageTitle)
	if spell.missing then
		return missingOutput(spell)
	end
	return Tooltip.render(spell)
end

function p.tooltip(frame)
	return p.renderTooltip(templateArgs(frame), currentTitleText())
end

function p.renderPageTooltip(args, pageTitle)
	local spell = p.resolve(args, pageTitle)
	if spell.missing then
		return ""
	end
	return Tooltip.render(spell)
end

function p.pageTooltip(frame)
	return p.renderPageTooltip(templateArgs(frame), currentTitleText())
end

return p
